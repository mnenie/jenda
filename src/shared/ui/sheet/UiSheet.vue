<script setup lang="ts">
import { computed, ref, useCssModule, useTemplateRef } from 'vue';
import { X } from 'lucide-vue-next';

interface SheetProps {
  overlay?: boolean;
  maxWidth?: number;
  transitionDuration?: number;
  minWidth?: number;
  overlayClickClose?: boolean;
}

const {
  overlay = true,
  maxWidth = 450,
  transitionDuration = 0.35,
  minWidth = 450,
  overlayClickClose = true
} = defineProps<SheetProps>();

const emit = defineEmits(['opened', 'closed']);

const showSheet = ref(false);
const sheetWidth = ref(0);
const translateValue = ref(100);
const overlayColor = ref('#0000004D');

const style = useCssModule();

const sideSheet = useTemplateRef<HTMLElement | null>('sideSheet');
const sideSheetHeader = useTemplateRef<HTMLElement | null>('sideSheetHeader');
const sideSheetMain = useTemplateRef<HTMLElement | null>('sideSheetMain');
const sideSheetFooter = useTemplateRef<HTMLElement | null>('sideSheetFooter');
const sideSheetContent = useTemplateRef<HTMLElement | null>('sideSheetContent');

document.addEventListener('keyup', (event: KeyboardEvent) => {
  if (event.key === 'Escape') {
    close();
  }
});

const sheetContentClasses = computed(() => {
  return {
    [style.sheet_content]: true,
    [style.sheet_content__fullscreen]: sheetWidth.value >= window.innerWidth
  };
});

const transitionDurationTime = computed(() => {
  return `${transitionDuration}s`;
});
const sheetWidthString = computed(() => {
  return sheetWidth.value && sheetWidth.value > 0 ? `${sheetWidth.value + 1}px` : 'auto';
});
const translateValueString = computed(() => {
  return `${translateValue.value}%`;
});
const maxWidthStyle = computed(() => {
  return `${maxWidth}px`;
});
const minWidthStyle = computed(() => {
  return `${minWidth}px`;
});

function open() {
  translateValue.value = 0;
  document.documentElement.style.overflowY = 'hidden';
  document.documentElement.style.overscrollBehavior = 'none';
  showSheet.value = true;
  emit('opened');
}
async function close() {
  showSheet.value = false;
  translateValue.value = 100;
  setTimeout(() => {
    document.documentElement.style.overflowY = 'auto';
    document.documentElement.style.overscrollBehavior = '';
    emit('closed');
  }, transitionDuration * 1000);
}

function clickOnOverlayHandler() {
  if (overlayClickClose) {
    close();
  }
}

defineExpose({ open, close });
</script>

<template>
  <Teleport to="body">
    <div ref="sideSheet" :class="$style.sheet" :aria-hidden="!showSheet" role="dialog">
      <Transition>
        <div v-show="overlay && showSheet" :class="$style.sheet_overlay" @click="clickOnOverlayHandler" />
      </Transition>
      <div ref="sideSheetContent" :class="sheetContentClasses">
        <header ref="sideSheetHeader" :class="$style.sheet_header">
          <slot name="header" />
          <X :size="17" :class="$style.header_icon" @click="close" />
        </header>
        <main ref="sideSheetMain" :class="$style.sheet_main">
          <slot />
        </main>
        <footer ref="sideSheetFooter" :class="[$style.sheet_footer, 'text-sm']">
          <slot name="footer" />
        </footer>
      </div>
    </div>
  </Teleport>
</template>

<style module lang="scss">
.sheet {
  z-index: 99999;
  display: flex;
  flex-direction: row;
  align-items: flex-start;
  justify-content: flex-end;

  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  transition: visibility v-bind('transitionDurationTime');

  &[aria-hidden='false'] {
    visibility: visible;
  }

  &[aria-hidden='true'] {
    visibility: hidden;
    pointer-events: none;
  }

  .sheet_overlay {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: -1;
    background: v-bind('overlayColor');
  }

  .sheet_content {
    display: flex;
    flex-direction: column;
    padding: 20px;
    min-width: v-bind('minWidthStyle');
    width: 100%;
    background: #ffffff;
    overflow-y: hidden;
    transform: translate3d(v-bind('translateValueString'), 0, 0);
    width: v-bind('sheetWidthString');
    max-width: v-bind('maxWidthStyle');
    height: 100%;
    box-sizing: border-box;
    pointer-events: all;

    .sheet_content__fullscreen {
      border-radius: 0;
    }

    &:not(.side-sheet__content--dragging) {
      transition: v-bind('transitionDurationTime') ease;
    }
  }

  .sheet_main {
    flex: 1;
    display: flex;
    flex-direction: column;
    overflow-y: auto;
    box-sizing: border-box;
    -webkit-overflow-scrolling: touch;
    touch-action: auto !important;
  }

  .sheet_footer {
    display: flex;
    margin-top: auto;
  }

  .sheet_footer:empty {
    display: none;
  }

  .sheet_header {
    display: flex;
    flex-direction: column;
    gap: 6px;
    padding: 5px 0px;
    margin-bottom: 20px;

    .header_icon {
      cursor: pointer;
      color: var(--zinc-600);
      position: absolute;
      right: 15px;
      top: 15px;
      transition: color 0.1s ease-in;

      &:hover {
        color: var(--zinc-900);
      }
    }
  }
}

:global(html.dark) {
  .sheet {
    .sheet_content {
      background-color: rgba(var(--zinc-rgb-800), 1);
    }
    .sheet_header {
      .header_icon {
        color: var(--zinc-400);
        &:hover {
          color: var(--zinc-300);
        }
      }
    }
  }
}
</style>
