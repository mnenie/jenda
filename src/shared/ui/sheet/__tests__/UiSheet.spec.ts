import { describe, expect, it } from 'vitest';
import { mount } from '@vue/test-utils';
import UiSheet from '../UiSheet.vue';

describe('UiSheet', () => {
  const wrapper = mount(UiSheet, {
    global: {
      stubs: {
        teleport: true
      }
    },
    props: {
      overlay: true,
      maxWidth: 450,
      transitionDuration: 0.35,
      overlayClickClose: true
    },
    slots: {
      header: 'sheet header',
      footer: 'sheet footer',
      default: 'test'
    }
  });

  it('should render correctly', async () => {
    // @ts-expect-error
    wrapper.vm.showSheet = true;
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render slots correctly', () => {
    expect(wrapper.find('[class*="sheet_header"]').text()).toContain('sheet header');
    expect(wrapper.find('[class*="sheet_main"]').text()).toContain('test');
    expect(wrapper.find('[class*="sheet_footer"]').text()).toContain('sheet footer');
  });

  it('should open and close the sheet', async () => {
    wrapper.vm.open();
    await wrapper.vm.$nextTick();
    expect(wrapper.find('[class*="sheet"]').attributes('aria-hidden')).toBe('false');

    wrapper.vm.close();
    await wrapper.vm.$nextTick();
    expect(wrapper.find('[class*="sheet"]').attributes('aria-hidden')).toBe('true');
  });
});
