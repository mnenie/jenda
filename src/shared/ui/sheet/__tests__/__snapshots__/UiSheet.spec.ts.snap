// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UiSheet > should render correctly 1`] = `
"<teleport-stub to="body">
  <div class="sheet" aria-hidden="false" role="dialog">
    <transition-stub appear="false" persisted="true" css="true">
      <div class="sheet_overlay"></div>
    </transition-stub>
    <div class="sheet_content">
      <header class="sheet_header">sheet header<svg xmlns="http://www.w3.org/2000/svg" width="17" height="17" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-xicon header_icon">
          <path d="M18 6 6 18"></path>
          <path d="m6 6 12 12"></path>
        </svg></header>
      <main class="sheet_main">test</main>
      <footer class="sheet_footer">sheet footer</footer>
    </div>
  </div>
</teleport-stub>"
`;
